src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -160,106 +200,92 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 additionalLocations));
         }
 
-        private static bool IsAccessedBeforeAssignment(
+        // Ensure that all usages of the pattern variable are
+        // in scope and definitely assigned after replacement.
+        private static bool CheckDefiniteAssignment(
             SemanticModel semanticModel,
             ISymbol localVariable,
             StatementSyntax declarationStatement,
             StatementSyntax targetStatement,
             BlockSyntax parentBlock,
-            CancellationToken cancellationToken)
+            bool isNegativeNullCheck)
         {
-            var isAssigned = false;
-            var isAccessedBeforeAssignment = false;
-
             var statements = parentBlock.Statements;
             var targetIndex = statements.IndexOf(targetStatement);
             var declarationIndex = statements.IndexOf(declarationStatement);
 
             // Since we're going to remove this declaration-statement,
             // we need to first ensure that it's not used up to the target statement.
-            for (var index = declarationIndex + 1; index < targetIndex; index++)
+            if (declarationIndex + 1 < targetIndex)
             {
-                CheckDefiniteAssignment(
-                    semanticModel, localVariable, statements[index],
-                    out isAssigned, out isAccessedBeforeAssignment,
-                    cancellationToken);
+                var dataFlow = semanticModel.AnalyzeDataFlow(
+                    statements[declarationIndex + 1], 
+                    statements[targetIndex - 1]);
 
-                if (isAssigned || isAccessedBeforeAssignment)
+                if (dataFlow.ReadInside.Contains(localVariable) || 
+                    dataFlow.WrittenInside.Contains(localVariable))
                 {
-                    return true;
+                    return false;
                 }
             }
 
             // In case of an if-statement, we need to check if the variable
-            // is being accessed before assignment in the else clause.
+            // is being accessed before assignment in the opposite branch.
             if (targetStatement.IsKind(SyntaxKind.IfStatement, out IfStatementSyntax ifStatement))
             {
-                CheckDefiniteAssignment(
-                    semanticModel, localVariable, ifStatement.Else,
-                    out isAssigned, out isAccessedBeforeAssignment,
-                    cancellationToken);
-
-                if (isAccessedBeforeAssignment)
+                var statement = isNegativeNullCheck ? ifStatement.Statement : ifStatement.Else?.Statement;
+                if (statement != null)
                 {
-                    return true;
-                }
+                    var dataFlow = semanticModel.AnalyzeDataFlow(statement);
+                    if (dataFlow.DataFlowsIn.Contains(localVariable))
```

Comment:
it woudl be helpful to have examples shown in the code of what this is catching.